pymeasure/instruments/teledyne/teledyneT3AFG.py

```diff
@@ -0,0 +1,131 @@
+
+from pymeasure.instruments import Instrument, Channel
+from pymeasure.instruments.validators import strict_range, strict_discrete_set
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class SignalChannel(Channel):
+
+    def get_process_generator(keyword, unit, type):
+        """Generate a get_process method searching for keyword, stripping unit"""
+        def selector(values):
+            if keyword in values:
+                try:
+                    return type(values[values.index(keyword)+1].strip(unit))
+                except (ValueError, IndexError):
+                    # Something went quite wrong if the keyword exists but the value doesn't
+                    return None
+            else:
+                # Wrong wavetype for this keyword
+                return None
+        return selector
+
+    output_enabled = Channel.control(
+        "C{ch}:OUTPut?",
+        "C{ch}:OUTPut %s",
+        """Control whether the channel output is enabled (boolean).""",
+        validator=strict_discrete_set,
+        map_values=True,
+        values={True: 'ON', False: 'OFF'},
+        # Replace ON and OFF with True and False in both get and set
+        get_process=lambda x: True if x[0].split(' ')[1] == 'ON' else False,
+    )
+
+    # TODO: Add other OUTPut related controls like Load and Polarity
+
+    wavetype = Channel.control(
+        "C{ch}:BSWV?",
+        "C{ch}:BSWV WVTP,%s",
+        """Control the type of waveform to be output.
+        Options are: {SINE, SQUARE, RAMP, PULSE, NOISE, ARB, DC, PRBS, IQ}
+        """,
+        validator=strict_discrete_set,
+        values=['SINE', 'SQUARE', 'RAMP', 'PULSE', 'NOISE', 'ARB', 'DC', 'PRBS', 'IQ'],
+        get_process=lambda x: x[1],
+    )
+
+    # TODO: Add frequency ranges per model?
+    frequency = Channel.control(
+        "C{ch}:BSWV?",
+        "C{ch}:BSWV FRQ,%g",
+        """Control the frequency of waveform to be output in Hertz.
+        Has no effect when WVTP is NOISE or DC.""",
+        validator=strict_range,
+        values=[0, 350e6],
+        get_process=get_process_generator('FRQ', 'HZ', float),
+    )
+
+    # TODO: Add range of outputs per model? Tricky without knowing offset
+    amplitude = Channel.control(
+        "C{ch}:BSWV?",
+        "C{ch}:BSWV AMP,%g",
+        """Control the amplitude of waveform to be output in volts peak-to-peak.
+        Has no effect when WVTP is NOISE or DC.""",
+        validator=strict_range,
+        values=[0, 5],
+        get_process=get_process_generator('AMP', 'V', float),
+    )
+
+    # TODO: Add range of offset per model? Tricky without knowing amplitude
+    offset = Channel.control(
+        "C{ch}:BSWV?",
+        "C{ch}:BSWV OFST,%g",
+        """Control the offset of waveform to be output in volts.
+        Has no effect when WVTP is NOISE.""",
+        validator=strict_range,
+        values=[0, 5],
+        get_process=get_process_generator('OFST', 'V', float),
+    )
```

Comment:
I'd add a comment to the docstrings, that the maximum value depends on the currently set offset/amplitude.
Does the instrument raise an error if the value is wrong? Could you read that error?

SCPI compliant devices allow to read the error state. You can call the method "check_errors" after a setter has set its value, if you add the option `check_set_errors=True`, see [documentation](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#checking-the-instrument-for-errors).

Reply:
I've added the check_set_errors flag and that appears to be working, though the -108.0, "Parameter not allowed" error appears after every command. The commands appear to work properly though, so I'm not sure what to make of the error appearing constantly.

Reply:
the `check_errors` sends "SYST:ERR?" (default SCPI command) to the device (several times, until no error is returned). Maybe that command is not right for your device.

Reply:
SYST:ERR? seems to be working fine, just every regular command I've tried seems to cause a parameter error. 

Unfortunately there isn't anything in the programming manual about variations on the default SCPI schema or examples including SYST:ERR? commands. There also isn't any detail on what specifically various errors are related to.

Reply:
For context, the debug log looks like this:
```
>>> t.ch_1.amplitude=1
DEBUG:WRITE:C1:BSWV AMP,1
DEBUG:WRITE:SYST:ERR?
DEBUG:TCPIP0::128.32.13.70::5023::SOCKET - reading 20480 bytes (last status <StatusCode.success_max_count_read: 1073676294>)
DEBUG:READ:-108,"Parameter not allowed"
ERROR:Teledyne T3AFG: -108.0, "Parameter not allowed"
DEBUG:WRITE:SYST:ERR?
DEBUG:TCPIP0::128.32.13.70::5023::SOCKET - reading 20480 bytes (last status <StatusCode.success_max_count_read: 1073676294>)
DEBUG:READ:0,"No error"
```
```
>>> t.ch_1.amplitude
DEBUG:WRITE:C1:BSWV?
DEBUG:TCPIP0::128.32.13.70::5023::SOCKET - reading 20480 bytes (last status <StatusCode.success_max_count_read: 1073676294>)
DEBUG:READ:C1:BSWV WVTP,SINE,FRQ,80000000HZ,PERI,1.25e-08S,AMP,1V,AMPVRMS,0.3535Vrms,MAX_OUTPUT_AMP,20V,OFST,5V,HLEV,5.5V,LLEV,4.5V,PHSE,0
1.0
```

Reply:
Interesting, that it always raises an error regarding invalid parameters.
Please add a note, probably to the instrument docstring, regarding this behaviour.

Reply:
Added a note, I'll come back and update if I ever figure out why that's happening

Reply:
Aha! I had a write termination of '\n' which worked initially with the USB interface, but does not work for the Ethernet interface that I'm testing with remotely. Removing the write termination solves this issue.

Reply:
You know how you can define the endings for different connection types?

Here is the corresponding documentation : https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#multiple-interfaces

Reply:
Ah, that's very handy! I'll add the ethernet interface that I am sure of for now. When I am back at work I'll test out all of my devices and I can go through and add more interface args for all of them.